### ISSUE TITLE
ListView Memory Leak

### ISSUE THREAD

**Issue by MelbourneDeveloper** [xamarin/Xamarin.Forms#2420]
### Description
When you set the ItemsSource on a ListView to a new list, the garbage collector should be able to finalise all the objects in the previous list and memory should be freed up. But, the objects are not getting garbage collected.

I don't know if this is a Xamarin Forms issue or a platform level issue.

**Note:**
_If I've missed something that I should have done to ensure the garbage collector can finalize the objects, please let me know_

### Steps to Reproduce
1. Clone https://ChristianFindlay@bitbucket.org/ChristianFindlay/xamarin-forms-scratch.git
2. Open the solution TestXamarinForms.sln in Test Apps
3. Run the UWP sample
4. Click on the ListView Memory Leak button
5. Debug->Windows-Show Diagnostics Tools
6. On the Diagnostics Tools click Take Snapshot
7. Click the "New Items" button 10 or so times
8. On the Diagnostics Tools click Take Snapshot
9. You will see two rows in the grid. Click in the cell of the column Objects (Diff) on the first hyperlink of the second row
10. Change Compare To -> "Snapshot #1"
11. Order the grid by "Object Type"
12. Scroll through the list to find TestXamarinFormsLibrary.Models.Fruit	
13. Notice that none of the objects have been garbage collected. It created five objects for each time you clicked the button
14. Click the "Return" button
15. Take another snapshot and see that the results are the same. Nothing was garbage collected
16. Open Task Manager and take note of the memory usage of the app
17. Click the New Items button 20-30 times. Notice that memory slowly creeps up and never goes back down even when you return to the main screen.

### Expected Behavior
Garbage collector finalises objects it has finished with

### Actual Behavior
Garbage collector cannot finalise the objects because they are connected to the root via Xamarin.Forms.TextCell.

### Basic Information
- Version with issue: 2.5.0.280555
- Last known good version: ?
- IDE: Visual Studio
- Platform Target Frameworks: 
  - UWP:  16299 
  - iOS:  Not tested but problem is probably the same
  - Android: Not tested but problem is probably the same

**Comment by PureWeen** [comment]
@MelbourneDeveloper  can you run this against the latest 3.0 pre 3 ?

I grabbed your sample and ran it against 3.0 pre 3 and when I did that I noticed in the diagnostic tools that only the current 5 fruits were still part of the active heap

**Comment by MelbourneDeveloper** [comment]
@PureWeen you are right. That's good news. It appears that this particular bug has been fixed in the latest release. I rechecked 2.5 on a different computer and can clearly see that it was a bug in that version. If you're interested, you should try with the previous version so you can see what I am talking about. 

However, I still suspect that there is more subtle leaking going on. If you click the New Items button a lot, you will notice that the process does gain a little more memory usage in Task Manager. But, it's nowhere near as bad as it was, and it does go down a little once the garbage has been collected. I'll do further investigation and log any issues that I come up with in a separate ticket.

I'm digging in to memory management at the moment because as I've pointed out #1897 , we have memory leak issues in our app. We have a long way to go because our app is made up of mostly dynamic XAML screens, so we have to go digging through the visual tree to unhook event handlers and so on to stop leaks. In this case, I didn't expect the leak to be so easy to find. I only started eyeballing ListView because the content is dynamically created via a DataTemplate and I figured this is where the leak was coming from. 

Long term, we need to resolve memory leak issues. I'm spending time determining where the issues are located: in the platform, Xamarin.Forms, or our code. The overall point I've been making in #1897 is: memory management is difficult. The request in #1897 was simply to allow for recycling of pages so as to circumvent the hornets nest of creating new objects, and then trying to break up the object mesh afterwards. 

The fact that the latest version of Xamarin Forms has moved forwards in the area of memory management is a positive sign that the XF team is paying attention to the issue and working toward making it easier. But, I'd still like to hear about other proposed memory management strategies in XF. Recycling pages was only one idea. We're only now embarking upon the exercise of ironing out the issues, so any advice would be helpful.

**Comment by MelbourneDeveloper** [comment]
@PureWeen , I am digging deep on this, so it's up to you if you want to close this ticket, or leave it open so that I can put new findings against this ticket.

**Comment by MelbourneDeveloper** [comment]
@PureWeen 

> Oh I just noticed you indicated this for UWP and across the platforms

I'm taking the rule of thumb that if the object left in the heap are not UWP specific classes, I'm assuming that this is a Xamarin.Forms wide issue and not a UWP specific issue. I can test on Android, but it's difficult and time consuming to test on iOS. However, it seems redundant to test on Android when the reference path is through the core objects - not UWP specific classes.

**Comment by MelbourneDeveloper** [comment]
@PureWeen ,

> I'll always just use the same instance of the Forms Page when I navigate and I notice that way I don't pay the layout costs. In my case I never have to navigate to it twice within the same stack though so it's a little different then 1897.

Are you able to elaborate on this? Do you have a sample of what you've done there?

**Comment by MelbourneDeveloper** [comment]
> nobody gets out alive

Ain't that the truth.

**Comment by PureWeen** [comment]
> Are you able to elaborate on this? Do you have a sample of what you've done there?

It's nothing too exciting.  I use ReactiveUI/Splat in all my application so I just register the page creations into Splat using this. 

```
dependencyResolver.RegisterLazySingleton
```
 Splat basically is just built around using a simple ServiceLocator type pattern for managing dependencies. It's really just a nice simple factory implementation that's fast and low weight.

When I navigate to a page I just ask Splat for the page that way it only news it up once and Splat internally just maintains a reference.  The only other *Magic*  I would say relies on Reactive concepts because of how easy it makes it to unwind events :-)  so I have Xamarin Forms Appearing/Disappearing tied into the VMS and then that's where I just wind up or wind down the events inside the VMS

I had lots of issues in SL as well with memory leaks back in the day :-) until I found Reactive which just made event leak management SO much easier to deal with.

I'm not sure if I have a useful sample of it anywhere though. I'll have to poke around. I have some samples but none of them have page reuse.

**Comment by MelbourneDeveloper** [comment]
@PureWeen I will research this further and see how it can be applied to XF.

Meantime, I'm just taking to document other leaks that I find from place to place. 

The WorkingWithNavigation sample is leaking a small amount. 

-Run the UWP sample
-Take a snapshot
-Click Next Page
-Click Next Page
-Click Return to Root Page
-Take a snapshot

You will notice that several objects are leaking. Most of these are common across the samples that I looked at. The worrying one here is FormsButton. It adds two for every time you go through the repro steps. That's the only non-UWP specific object I noticed. However, the other ones are the ones eating up memory.
![image](https://user-images.githubusercontent.com/16697547/38715329-d2eeb146-3f1d-11e8-9cc9-169c26ea2fc4.png)

Note: I upgraded to 2.5.1.444934 but I can't get V3 (pre) working with any samples at the moment.
